projects/cdk/directives/auto-focus/autofocus.directive.ts

```diff
@@ -32,57 +32,85 @@ export class TuiAutoFocusDirective implements AfterViewInit {
     autoFocus = true;
 
     constructor(
-        @Inject(INJECTOR) private readonly injector: Injector,
-        @Inject(ElementRef) private readonly elementRef: ElementRef<HTMLElement>,
         @Optional()
         @Self()
         @Inject(TUI_FOCUSABLE_ITEM_ACCESSOR)
         private readonly tuiFocusableComponent: TuiFocusableElementAccessor | null,
         @Inject(TUI_IS_IOS) private readonly isIos: boolean,
         @Inject(ANIMATION_FRAME) private readonly animationFrame$: Observable<number>,
+        @Inject(NgZone)
+        private readonly ngZone: NgZone,
+        @Inject(ElementRef)
+        private readonly elementRef: ElementRef<HTMLElement>,
+        @Inject(Renderer2)
+        private readonly renderer: Renderer2,
     ) {}
 
     ngAfterViewInit() {
         if (!this.autoFocus) {
             return;
         }
 
-        const element =
+        if (this.isIos && this.isTextFieldElement()) {
+            this.ngZone.runOutsideAngular(() => this.iosWebkitAutofocus());
+        } else {
+            race(
+                timer(TIMEOUT),
+                this.animationFrame$.pipe(
+                    throttleTime(POLLING_TIME),
+                    map(() => this.element.closest(NG_ANIMATION_SELECTOR)),
+                    skipWhile(Boolean),
+                    take(1),
+                ),
+            ).subscribe(() => setNativeFocused(this.element));
+        }
+    }
+
+    private get element(): TuiNativeFocusableElement {
+        return (
             this.tuiFocusableComponent?.nativeFocusableElement ||
-            this.elementRef.nativeElement;
+            this.elementRef.nativeElement
+        );
+    }
 
-        if (this.isIos) {
-            veryVerySadIosFix(this.injector);
+    private isTextFieldElement(): boolean {
```

Comment:
This can be a case insensitive one-liner `element.matches('input, textarea')`

Reply:
fixed